File: java-practice/src/main/java/practice/util/TargetResolver.java
package main.java.practice.util;

import java.io.File;
import java.net.URL;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * TargetResolver automatically finds the latest implementation of a challenge.
 *
 * HOW IT WORKS:
 * -------------
 * 1. Scans the classpath for .class files matching the base name
 * 2. Extracts version numbers from filenames (LargestSum2 -> version 2)
 * 3. Returns the highest numbered version
 * 4. Allows override via PRACTICE_TARGET environment variable
 *
 * For a challenge named "LargestSum", it will look for:
 *   - LargestSum.java (attempt 0 - no number)
 *   - LargestSum2.java (attempt 2)
 *   - LargestSum3.java (attempt 3)
 *   etc.
 *
 * By default, it returns the highest numbered attempt.
 * You can override with environment variable PRACTICE_TARGET=LargestSum2
 *
 * EXAMPLE USAGE:
 * --------------
 *   // In your test class:
 *   LargestSumSolution solution = TargetResolver.resolve("practice.largestsum", "LargestSum");
 *   // Returns instance of LargestSum3 if that's the highest version
 */
public class TargetResolver {

    // Environment variable name for overriding which version to test
    private static final String ENV_OVERRIDE = "PRACTICE_TARGET";

    /**
     * Find and instantiate the target implementation class.
     *
     * PROCESS:
     * --------
     * 1. Check for PRACTICE_TARGET environment variable
     * 2. If set, use that specific class
     * 3. If not set, find the latest (highest numbered) implementation
     * 4. Load the class using reflection
     * 5. Create and return a new instance
     *
     * @param basePackage The package containing the challenge (e.g., "practice.largestsum")
     * @param baseName The base class name (e.g., "LargestSum")
     * @return Instance of the target class (cast to your interface type)
     * @throws Exception if class not found or cannot be instantiated
     */
    @SuppressWarnings("unchecked")
    public static <T> T resolve(String basePackage, String baseName) throws Exception {
        // Check if user wants to test a specific version via environment variable
        String envTarget = System.getenv(ENV_OVERRIDE);

        String targetClassName;
        if (envTarget != null && !envTarget.isEmpty()) {
            // Use environment variable override (e.g., PRACTICE_TARGET=LargestSum2)
            targetClassName = basePackage + "." + envTarget;
            System.out.println("🎯 Using PRACTICE_TARGET override: " + targetClassName);
        } else {
            // Find the latest attempt automatically
            targetClassName = findLatestAttempt(basePackage, baseName);
            System.out.println("🎯 Testing latest attempt: " + targetClassName);
        }

        // Use reflection to load the class and create an instance
        Class<T> clazz = (Class<T>) Class.forName(targetClassName);
        return clazz.getDeclaredConstructor().newInstance();
    }

    /**
     * Find the latest numbered attempt of a challenge.
     *
     * ALGORITHM:
     * ----------
     * 1. Convert package name to file path (. -> /)
     * 2. Search classpath for the package directory
     * 3. List all .class files that start with baseName
     * 4. Extract version numbers from filenames
     * 5. Return the class with the highest version number
     *
     * EXAMPLE:
     * For baseName="LargestSum", finds: LargestSum.class, LargestSum2.class, LargestSum3.class
     * Returns: "practice.largestsum.LargestSum3"
     *
     * @param basePackage Package to search (e.g., "practice.largestsum")
     * @param baseName Base class name (e.g., "LargestSum")
     * @return Full class name of latest version (e.g., "practice.largestsum.LargestSum3")
     * @throws ClassNotFoundException if no matching classes found
     */
    private static String findLatestAttempt(String basePackage, String baseName) throws Exception {
        // List to store all found class names
        List<String> foundClasses = new ArrayList<>();

        // Convert package name to path (practice.largestsum -> practice/largestsum)
        String packagePath = basePackage.replace('.', '/');

        // Get the classloader to search for compiled .class files
        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
        Enumeration<URL> resources = classLoader.getResources(packagePath);

        // Regex pattern to match: LargestSum, LargestSum2, LargestSum3, etc.
        // The (\\d*) captures optional digits after the base name
        Pattern pattern = Pattern.compile(baseName + "(\\d*)");

        // Search through all locations where this package appears
        while (resources.hasMoreElements()) {
            URL resource = resources.nextElement();
            File directory = new File(resource.toURI());

            if (directory.exists() && directory.isDirectory()) {
                // Filter files: must end with .class and start with baseName
                File[] files = directory.listFiles((dir, name) ->
                    name.endsWith(".class") && name.startsWith(baseName));

                if (files != null) {
                    for (File file : files) {
                        // Remove .class extension to get class name
                        String className = file.getName().replace(".class", "");
                        Matcher matcher = pattern.matcher(className);
                        if (matcher.matches()) {
                            foundClasses.add(className);
                        }
                    }
                }
            }
        }

        // Throw error if no implementations found
        if (foundClasses.isEmpty()) {
            throw new ClassNotFoundException(
                "No implementation found for " + baseName + " in package " + basePackage);
        }

        // Sort classes by version number and get the highest
        String latest = foundClasses.stream()
            .max(Comparator.comparingInt(TargetResolver::extractAttemptNumber))
            .orElseThrow();

        // Return fully qualified class name
        return basePackage + "." + latest;
    }

    /**
     * Extract the attempt number from a class name.
     *
     * EXAMPLES:
     * ---------
     * LargestSum   -> 0 (no number, so it's attempt 0/baseline)
     * LargestSum2  -> 2
     * LargestSum10 -> 10
     *
     * HOW IT WORKS:
     * Uses regex to find digits at the end of the class name.
     * If no digits found, returns 0 (making baseline version lowest priority).
     *
     * @param className Simple class name (e.g., "LargestSum2")
     * @return Version number (2 for LargestSum2, 0 for LargestSum)
     */
    private static int extractAttemptNumber(String className) {
        // Regex: \D+ (one or more non-digits) followed by (\d+)$ (digits at end)
        Pattern pattern = Pattern.compile("\\D+(\\d+)$");
        Matcher matcher = pattern.matcher(className);

        if (matcher.find()) {
            // Extract and parse the digit group
            return Integer.parseInt(matcher.group(1));
        }
        // No number found - this is the base version (attempt 0)
        return 0;
    }
}

File: js-practice/utils/targetResolver.js
// Import Node.js modules for file system operations
const fs = require("fs"); // File system - read files and directories
const path = require("path"); // Path manipulation - join paths correctly

/**
 * TargetResolver - Automatically finds the latest implementation of a challenge
 *
 * HOW IT WORKS:
 * -------------
 * 1. Scans the practice directory for .js files matching the base name
 * 2. Extracts version numbers from filenames (largestSum2.js -> version 2)
 * 3. Returns the file with the highest version number
 * 4. Allows override via PRACTICE_TARGET environment variable
 *
 * For a challenge named "largestSum", it will look for:
 *   - largestSum.js (attempt 0 - no number)
 *   - largestSum2.js (attempt 2)
 *   - largestSum3.js (attempt 3)
 *   etc.
 *
 * By default, returns the highest numbered attempt.
 * Override with environment variable: PRACTICE_TARGET=largestSum2
 *
 * EXAMPLE USAGE:
 * --------------
 *   // In your test file:
 *   const solution = TargetResolver.resolve('largestSum');
 *   // Returns exports from largestSum3.js if that's the highest version
 */
class TargetResolver {
  /**
   * Find and require the target implementation module.
   *
   * PROCESS:
   * --------
   * 1. Check for PRACTICE_TARGET environment variable
   * 2. If set, use that specific file
   * 3. If not set, find the latest (highest numbered) file
   * 4. Require (import) that file
   * 5. Return its exports
   *
   * @param {string} challengeName - Base name of the challenge (e.g., "largestSum")
   * @param {string} practiceDir - Directory containing practice files (default: "../practice")
   * @returns {*} The exported module from the target file (e.g., { bigSum: function })
   */
  static resolve(challengeName, practiceDir = path.join(__dirname, "../practice")) {
    // Check if user wants to test a specific version via environment variable
    const envTarget = process.env.PRACTICE_TARGET;

    let targetFile;
    if (envTarget) {
      // Use environment variable override (e.g., PRACTICE_TARGET=largestSum2)
      targetFile = `${envTarget}.js`;
      console.log(`🎯 Using PRACTICE_TARGET override: ${targetFile}`);
    } else {
      // Find the latest attempt automatically
      targetFile = this.findLatestAttempt(challengeName, practiceDir);
      console.log(`🎯 Testing latest attempt: ${targetFile}`);
    }

    // Build full path to the target file
    const targetPath = path.join(practiceDir, targetFile);

    // Verify file exists before trying to require it
    if (!fs.existsSync(targetPath)) {
      throw new Error(`Target file not found: ${targetPath}`);
    }

    // Import and return the module (what it exports)
    return require(targetPath);
  }

  /**
   * Find the latest numbered attempt of a challenge.
   *
   * ALGORITHM:
   * ----------
   * 1. Read all files in the practice directory
   * 2. Filter files that match pattern: baseName + optional number + .js
   * 3. Extract version numbers (no number = 0)
   * 4. Sort by version number (descending)
   * 5. Return the highest version
   *
   * EXAMPLE:
   * For baseName="largestSum", finds: largestSum.js, largestSum2.js, largestSum3.js
   * Returns: "largestSum3.js"
   *
   * @param {string} baseName - Base name (e.g., "largestSum")
   * @param {string} dir - Directory to search
   * @returns {string} Filename of latest attempt (e.g., "largestSum3.js")
   */
  static findLatestAttempt(baseName, dir) {
    // Verify directory exists
    if (!fs.existsSync(dir)) {
      throw new Error(`Practice directory not found: ${dir}`);
    }

    // Read all files in directory
    const files = fs.readdirSync(dir);

    // Create regex pattern to match: baseName + optional digits + .js
    // Example: /^largestSum(\d*)\.js$/
    // This matches: largestSum.js, largestSum2.js, largestSum10.js
    const pattern = new RegExp(`^${baseName}(\\d*)\\.js$`);

    // Filter and transform matching files
    const matches = files
      .filter((file) => pattern.test(file)) // Keep only matching files
      .map((file) => {
        const match = file.match(pattern);
        // Extract number from filename (empty string -> 0, "2" -> 2)
        const attemptNum = match[1] ? parseInt(match[1], 10) : 0;
        return { file, attemptNum };
      })
      .sort((a, b) => b.attemptNum - a.attemptNum); // Sort descending (highest first)

    // Throw error if no matching files found
    if (matches.length === 0) {
      throw new Error(`No implementation found for ${baseName} in ${dir}`);
    }

    // Return the filename with highest version number
    return matches[0].file;
  }
}

// Export the TargetResolver class so other files can use it
// Usage: const TargetResolver = require('./targetResolver');
module.exports = TargetResolver;
